flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java

```diff
@@ -169,14 +171,44 @@ private void captureNewlyAddedTables() {
             // check whether we got newly added tables
             try (JdbcConnection jdbc = openJdbcConnection(sourceConfig)) {
                 final List<TableId> newlyAddedTables = discoverCapturedTables(jdbc, sourceConfig);
+                List<TableId> prevTables = new LinkedList<>(alreadyProcessedTables);
+                prevTables.addAll(remainingTables);
+                prevTables.removeAll(newlyAddedTables);
+
                 newlyAddedTables.removeAll(alreadyProcessedTables);
                 newlyAddedTables.removeAll(remainingTables);
-                if (!newlyAddedTables.isEmpty()) {
+                if (!newlyAddedTables.isEmpty() || !prevTables.isEmpty()) {
                     // if job is still in snapshot reading phase, directly add all newly added
                     // tables
                     LOG.info("Found newly added tables, start capture newly added tables process");
+                    Tables.TableFilter tableFilter =
+                            sourceConfig.getTableFilters().dataCollectionFilter();
+
+                    // remove unassigned tables/splits if it does not satisfy new table filter
+                    remainingTables.removeIf(tableId -> !tableFilter.isIncluded(tableId));
+
+                    List<String> splitsToRemove = new LinkedList<>();
+                    for (Map.Entry<String, MySqlSnapshotSplit> splitEntry :
+                            assignedSplits.entrySet()) {
+                        if (!tableFilter.isIncluded(splitEntry.getValue().getTableId())) {
+                            splitsToRemove.add(splitEntry.getKey());
+                        }
+                    }
+
+                    splitsToRemove.forEach(assignedSplits.keySet()::remove);
+                    splitsToRemove.forEach(splitFinishedOffsets.keySet()::remove);
+
+                    remainingSplits.removeIf(split -> !tableFilter.isIncluded(split.getTableId()));
+                    // add new tables
                     remainingTables.addAll(newlyAddedTables);
-                    if (isAssigningFinished(assignerStatus)) {
+
+                    // remove unmatched tables from already processed tables since the abnormal
+                    // tables might be added back in the future
+                    alreadyProcessedTables.removeIf(tableId -> !tableFilter.isIncluded(tableId));
+
+                    // only suspend the assigner if there's new tables added and assigner has
+                    // finished
+                    if (isAssigningFinished(assignerStatus) && !newlyAddedTables.isEmpty()) {
```

Comment:
`!newlyAddedTables.isEmpty()` is called twice, can we split the removed and added logic to make it more clear?

Reply:
Changed the code to remove tables first and then add new tables.